### PR TITLE
Adding DDP to the repo for multi-GPU training

### PR DESCRIPTION
**Pull request by vratins**
Multi-GPU flow matching training via torchrun

New `src/distributed.py` (shared DDP helpers) + DDP wiring in `train.py`. No CLI flag — `torchrun` env vars are the only switch, single-GPU runs unchanged. Verified on 2×H100. tests are CPU-only for CI for now.

Updated README.

### CODE REVIEW

File: scripts/train.py [hunk 1]
```diff
@@ -763,11 +811,17 @@ def train_epoch(
             skipped_batches += 1
             continue
```
[the remark this answers is not in the document]
Comment (vratins) [reply scripts/train.py:812]: By construction the data will not have any batches with 0 water and that will be caught early, and this is more of a debug

File: src/distributed.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+def is_main_process(rank: int) -> bool:
+    """True on the single rank that owns disk IO and W&B logging."""
+    return rank == 0
+
+
+def ddp_barrier() -> None:
+    """Block until every rank arrives. No-op when not distributed."""
+    if ddp_is_active():
+        dist.barrier()
```
Comment (marcuscollins) [inline src/distributed.py:83]: I thought dist.barrier did this check already? Or rather, I thought it didn't matter (i.e., DDP still works with a world size == 1)

File: README.md [hunk 3]
```diff
@@ -300,6 +302,27 @@ uv run python -m scripts.train \
     --processed_dir ~/flow_cache/
 ```
 
+### Multi-GPU Training (DDP)
+
+No DDP flag — `torchrun`'s env vars are the only switch; a plain
```
Comment (marcuscollins) [inline README.md]: This is cryptic AI-speak. If someone doesn't know what `torchrun` is, this doesn't help them. Rewrite declaratively, just tell people what to do if they want to use more than one GPU.

File: README.md [hunk 4]
```diff
@@ -300,6 +302,27 @@ uv run python -m scripts.train \
...
+
+```bash
+uv run torchrun --nproc_per_node=4 -m scripts.train \
+    --train_list splits/train_list_0.95.txt \
+    --val_list splits/valid_list_0.05.txt \
+    --encoder_type gvp \
+    --batch_size 4  # per rank -> effective 16
+```
+
+- Each rank trains on a disjoint `DistributedSampler` shard, reshuffled per epoch.
```
Comment (marcuscollins) [inline README.md]: Remove this line.

File: scripts/train.py [hunk 5]
```diff
@@ -1085,37 +1241,56 @@ def main():
...
             num_graphs = int(batch["protein"].batch.max().item()) + 1
             t = torch.zeros(num_graphs, device=device)
-            v_out = model(batch, t)
+            v_out = raw_model(batch, t)
             logger.info(f"Forward pass successful! Output shape: {v_out.shape}")
             logger.info(f"Output stats: mean={v_out.mean():.4f}, std={v_out.std():.4f}")
             if v_out.std() < 1e-6:
                 logger.warning("Model output is constant! This indicates a problem.")
-        model.train()
+        raw_model.train()
```
Comment (marcuscollins) [inline scripts/train.py:1293]: I think I would still set this on `model`, not just `raw_model`. I don't remember well enough whether there are other things in the DDP object that might need setting.

File: scripts/train.py [hunk 6]
```diff
@@ -1085,37 +1241,56 @@ def main():
...
 
     flow_matcher = FlowMatcher(
         model=model,
```
Comment (marcuscollins) [inline scripts/train.py:1296]: Does the FlowMatcher need/have updates to understand the DDP object?

File: scripts/train.py [hunk 7]
```diff
@@ -733,17 +756,42 @@ def run_eval_sampling(
...
-            "eval/avg_precision": np.mean([r["precision"] for r in results]),
-            "eval/avg_recall": np.mean([r["recall"] for r in results]),
-            "eval/avg_f1": np.mean([r["f1"] for r in results]),
-            "eval/avg_auc_pr": np.mean([r["auc_pr"] for r in results]),
-        }
-        wandb.log(avg_metrics, step=global_step)
-        return avg_metrics
-    return {}
+    # Every rank must reach this, even one whose stride was all zero-water graphs.
+    avg_metrics, _ = all_reduce_means(
```
Comment (marcuscollins) [inline scripts/train.py:760]: There is an AVG reduce operation in torch. Why not use it directly with `dist.reduce`?

File: scripts/train.py [hunk 8]
```diff
@@ -770,11 +818,17 @@ def train_epoch(
...
-            batch,
-            use_self_conditioning=args.use_self_cond,
-            accumulation_steps=args.grad_accum_steps,
+        # Suppress the gradient all-reduce on micro-steps that are not followed by
+        # an optimizer.step(), keeping comms at one all-reduce per optimizer step.
+        no_sync = ddp_is_active() and not _needs_grad_sync(
+            step, len(train_loader), args.grad_accum_steps
         )
+        with flow_matcher.model.no_sync() if no_sync else contextlib.nullcontext():
+            metrics = flow_matcher.training_step(
```
Comment (marcuscollins) [inline scripts/train.py:818]: This feels weird to me... don't you have to set a barrier somewhere and do some reduce operation? It's been a while since I've looked at the low-level DDP package, is the DDP object itself handling the synchronization and reduction of gradients?

File: src/distributed.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    keys = list(sums)
+    totals = torch.tensor(
+        [float(sums[k]) for k in keys] + [float(count)],
+        dtype=torch.float64,
+        device=device,
+    )
+    if ddp_is_active():
+        dist.all_reduce(totals, op=dist.ReduceOp.SUM)
+
+    total_count = int(totals[-1].item())
```
Comment (marcuscollins) [inline src/distributed.py]: Use `total_count = totals[-1].item()` here, so you can re-use it as the denominator below, and then return `int(total_count)` in the return line.

File: src/distributed.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    if not ddp_is_active():
+        work()
+        return None
+
+    rank = int(os.environ["RANK"])
+    world = int(os.environ["WORLD_SIZE"])
+    host = os.environ["MASTER_ADDR"]
+    port = int(os.environ["MASTER_PORT"])
+    # Client of torchrun's agent store (is_master=False) -- the agent already owns
+    # MASTER_PORT. The timeout covers a cold build, which can take hours.
```
Comment (marcuscollins) [inline src/distributed.py]: I don't understand what you're building that could take up to 24 hours? If it is the data set, you shouldn't be doing that every time you run anyway. Anyway it seems like this kind of thing is better handled in your training script, protecting it in a conditional that runs only on rank 0, and using a barrier to prevent the other processes from proceeding until all ranks (in this case only 0) are done.

File: scripts/train.py [hunk 11]
```diff
@@ -847,6 +901,14 @@ def train_epoch(
 
     final_global_step = (epoch - 1) * len(train_loader) + len(train_loader) - 1
 
+    # One collective per epoch so metrics cover every rank's shard. Must run before
```
Comment (marcuscollins) [inline scripts/train.py]: Do you mean "one collection"? I don't understand this comment at all.

File: scripts/train.py [hunk 12]
```diff
@@ -847,6 +901,14 @@ def train_epoch(
...
+    # the zero-batch check, or a rank that skipped everything would never enter it.
+    train_metrics, processed_batches = all_reduce_means(
+        {"train/epoch_loss": total_loss, "train/epoch_rmsd": total_rmsd},
+        processed_batches,
+        _collective_device(args),
```
Comment (marcuscollins) [inline scripts/train.py]: So, if I understand correctly, _collective_device should return the current rank device this particular instance of the script is running with. This feels clunky and error prone. Usually you want to set the device for your process, and then use that, so that the device is the same as local_rank. I thought torchrun did this by default by setting CUDA_VISIBLE_DEVICES to local_rank when it spawns the subprocesses, so that all you'd need to do here is pass local_rank.

File: src/distributed.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+
+    Returns:
+        The store, for `setup_distributed(store=...)`; None when not distributed
+        (in which case `work` simply runs inline).
+    """
+    if not ddp_is_active():
+        work()
+        return None
+
+    rank = int(os.environ["RANK"])
```
Comment (marcuscollins) [inline src/distributed.py:114]: why not use `dist.get_rank()` etc. here?

File: scripts/train.py [hunk 14]
```diff
@@ -992,27 +1051,121 @@ def build_scheduler(optimizer, args):
...
+        shutil.rmtree(tmpdir, ignore_errors=True)
+
+
 def main():
     """Run the full training pipeline."""
     args = parse_args()
+
+    # Build the cache before the NCCL group exists: a cold build can take hours and
+    # would trip a GPU-collective timeout. The store is reused as NCCL's rendezvous.
+    store = run_once_on_main(lambda: build_cache(args), key="wf_cache_ready")
```
Comment (marcuscollins) [inline scripts/train.py:1130]: If this can take hours, and is the same for many training runs, you should not be running it as part of this script at all. Instead, build it once if that is possible, then load the data. Then you can avoid needing the store and any associated timeouts.

File: scripts/train.py [hunk 15]
```diff
@@ -992,27 +1051,121 @@ def build_scheduler(optimizer, args):
...
+    hence race-free. A warm cache is a fast no-op; a cold build is parallelized
+    across CPU cores over disjoint key shards.
+    """
+    dataset_kwargs, _, _ = _build_dataset_config(args)
+    ids = set()
+    for lst in (args.train_list, args.val_list):
+        with open(lst) as f:
+            ids.update(line.strip() for line in f if line.strip())
+    sorted_ids = sorted(ids)
+    if not sorted_ids:
```
Comment (marcuscollins) [inline scripts/train.py]: always bail out as soon as you can. You can check if `ids` is empty before sorting.

File: scripts/train.py [hunk 16]
```diff
@@ -992,27 +1051,121 @@ def build_scheduler(optimizer, args):
...
+    Build the geometry cache for the train+val lists as the sole writer.
+
+    Preprocessing is CPU/PyMOL only, so this runs before the DDP group exists --
+    hence race-free. A warm cache is a fast no-op; a cold build is parallelized
+    across CPU cores over disjoint key shards.
+    """
+    dataset_kwargs, _, _ = _build_dataset_config(args)
+    ids = set()
+    for lst in (args.train_list, args.val_list):
+        with open(lst) as f:
```
Comment (marcuscollins) [inline scripts/train.py:1075]: You can also use `for line in open(list):`, the open method provides the correct context that will close the file when the iterator is exhausted.

File: scripts/train.py [hunk 17]
```diff
@@ -992,27 +1051,121 @@ def build_scheduler(optimizer, args):
...
+            ids.update(line.strip() for line in f if line.strip())
+    sorted_ids = sorted(ids)
+    if not sorted_ids:
+        return
+
+    tmpdir = Path(tempfile.mkdtemp(prefix="wf_build_"))
+    try:
```
Comment (marcuscollins) [inline scripts/train.py:1082]: There's no except block? I feel like you should at least log the error.

File: scripts/train.py [hunk 18]
```diff
@@ -992,27 +1051,121 @@ def build_scheduler(optimizer, args):
...
+        union.write_text("\n".join(sorted_ids) + "\n")
+        # Parse-only probe to find which entries still need building.
+        probe = ProteinWaterDataset(
+            pdb_list_file=str(union),
+            processed_dir=args.processed_dir,
+            preprocess=False,
+            **dataset_kwargs,
+        )
+        missing = [
+            entry["cache_key"]
```
Comment (marcuscollins) [inline scripts/train.py]: There are fewer cache keys than entries, right? If so, make a set of the keys and only check for each file once, rather than for every entry.

File: scripts/train.py [hunk 19]
```diff
@@ -992,27 +1051,121 @@ def build_scheduler(optimizer, args):
...
+            if not (probe.geometry_dir / f"{entry['cache_key']}.pt").is_file()
+        ]
+        if not missing:
+            return  # warm cache: nothing to build
+
+        logger.info(f"build_cache: preprocessing {len(missing)} missing entries")
+        n_shards = max(1, min(len(missing), os.cpu_count() or 1))
+        if n_shards == 1:
+            _build_cache_shard(str(union), args.processed_dir, dataset_kwargs)
+            return
```
Comment (marcuscollins) [inline scripts/train.py]: This seems unnecessary. I don't think there's that much overhead if you spawn another process for this task, even if there's just one shard.

File: scripts/train.py [hunk 20]
```diff
@@ -992,27 +1051,121 @@ def build_scheduler(optimizer, args):
...
+        shard_files = []
+        for i in range(n_shards):
+            shard = tmpdir / f"shard_{i}.txt"
+            shard.write_text("\n".join(missing[i::n_shards]) + "\n")
+            shard_files.append(str(shard))
+        # spawn (not fork): safe alongside PyMOL's C extension and any threads.
+        ctx = mp.get_context("spawn")
```
Comment (marcuscollins) [inline scripts/train.py:1108]: In the future I encourage you to use higher-level interfaces like `joblib` which makes this kind of thing trivially easy. Here it isn't too complex, since _build_cache_shard only raises an error or returns None, but if you care about the output `joblib` makes life much simpler.

File: scripts/train.py [hunk 21]
```diff
@@ -992,27 +1051,121 @@ def build_scheduler(optimizer, args):
...
+
+    # Build the cache before the NCCL group exists: a cold build can take hours and
+    # would trip a GPU-collective timeout. The store is reused as NCCL's rendezvous.
+    store = run_once_on_main(lambda: build_cache(args), key="wf_cache_ready")
+
+    # Under torchrun each rank binds its own GPU; a plain launch yields (0, 0, 1).
+    rank, local_rank, world_size = setup_distributed(store=store)
+    main_proc = is_main_process(rank)
+    if ddp_is_active():
+        args.device = f"cuda:{local_rank}"
```
Comment (marcuscollins) [inline scripts/train.py]: Okay, regarding an above comment, don't add this complexity. Just use `local_rank` when you need to figure out what device to send a tensor to. You don't need to append it to `args` for this.

**Review by marcuscollins — APPROVED** [review]
The code looks technically correct to me, but there are a lot of things I would change or look into. Given time constraints I'm approving, but please look over the comments and update what you can.

File: src/distributed.py (hunk 2 above)
Comment (vratins) [reply src/distributed.py:83]: For the barrier to work you need a process group, and dist.barrier() would crash with a single GPU run

File: README.md (hunk 3 above)
Comment (vratins) [reply README.md]: Done

File: README.md (hunk 4 above)
Comment (vratins) [reply README.md]: Done

File: scripts/train.py (hunk 5 above)
Comment (vratins) [reply scripts/train.py:1293]: I checked, and nothing else needs setting wit DDP, and setting this on model made the block look inconsistent, hence went with raw_model since that is where the eval flags are being set anyway

File: scripts/train.py (hunk 6 above)
Comment (vratins) [reply scripts/train.py:1296]: FlowMatcher is handed the wrapped model, and custom attributes are handled in flow.py, checked this.

File: scripts/train.py (hunk 8 above)
Comment (vratins) [reply scripts/train.py:818]: Yes, DDP does that itself, it attaches a hook to every parameter, and during backward(), DDP handles the reduce. No need for manual barrier/reduce for gradients.

File: scripts/train.py (hunk 7 above)
Comment (vratins) [reply scripts/train.py:760]: AVG uses the mean, but not weighted by the number of samples on a GPU. Technically the sum and then divide we are doing is equivalent to all our training cases since this is an issue only when we have imbalanced batch sizes across different GPUs, and that would be when we skip a batch due to no waters -- but this is taken care of in pre-processing, before training. But I kept sum-then-divide because it stays correct if counts ever diverge as a guard.

File: src/distributed.py (hunk 9 above)
Comment (vratins) [reply src/distributed.py]: Yep, done.

File: scripts/train.py (hunk 11 above)
Comment (vratins) [reply scripts/train.py]: Meant the all-reduce operation. Re-worded.

File: scripts/train.py (hunk 15 above)
Comment (vratins) [reply scripts/train.py]: Done.

File: src/distributed.py (hunk 13 above)
Comment (vratins) [reply src/distributed.py:114]: `run_once_on_main` runs before `setup_distributed/init_process_group`, so there's no process group yet and `dist.get_rank()` would raise an error. Might change this while addressing the comments about building the dataset before any of this

File: src/distributed.py (hunk 10 above)
Comment (vratins) [reply src/distributed.py]: A 24 hour run would only be if we are running it from scratch, but the code also goes through any missing entries from given split -- clarified comments towards that.

File: scripts/train.py (hunk 12 above)
Comment (vratins) [reply scripts/train.py]: torchrun doesn't set CUDA_VISIBLE_DEVICES, it only sets LOCAL_RANK, and we call torch.cuda.set_device(local_rank) ourselves, so each process sees all GPUs and must target cuda:local_rank (which args.device holds). But yes, the helper is redundant, so I unified device handling (addressing line 1140 too)

File: scripts/train.py (hunk 14 above)
Comment (vratins) [reply scripts/train.py:1130]: Clarified that comment; I think we need the store to safeguard missing caches

File: scripts/train.py (hunk 16 above)
Comment (vratins) [reply scripts/train.py:1075]: I chose to keep it this way since I was getting ruff flags (SIM115)

File: scripts/train.py (hunk 17 above)
Comment (vratins) [reply scripts/train.py:1082]: Added one.

File: scripts/train.py (hunk 18 above)
Comment (vratins) [reply scripts/train.py]: Done

File: scripts/train.py (hunk 19 above)
Comment (vratins) [reply scripts/train.py]: Addressed

File: scripts/train.py (hunk 20 above)
Comment (vratins) [reply scripts/train.py:1108]: `_build_cache_shard` returns None/raises, so there's no output to collect hence went with mp.Pool. Will explore joblib in the future!

File: scripts/train.py (hunk 21 above)
Comment (vratins) [reply scripts/train.py]: Yep, addressed. main() computes once device that's threaded through.
